### ISSUE TITLE
save to disk exception

### ISSUE THREAD

**Issue by ProfLogic** [BlueWallet/BlueWallet#3638]
When creating or importing a (new) Lightning wallet, the following error message shows up:

> save to disk exception
>
> code 9
> api-level 28
> message Failed to load generated key pair from keystore

Funds can be sent and received, but on the next start of the App, the wallet is not shown anymore and lost (forever).

**Comment by Overtorment** [comment]
this sounds like an issue with keystore (a secure phone storage). have you tried deleting and reinstalling bluewallet?

also, what android ver is that?

**Comment by ProfLogic** [comment]
Deleting and reinstalling BlueWallet did not change anything. Also installing BlueWallet on an equal device leads to the same behavior. The Android version is 9 (API 28, as written in the original post). Here LineageOS 16.0 is used. However, on other rooted smartphones with LineageOS the App works well.

**Comment by Overtorment** [comment]
could it be that some permission is disabled for the app? like the permission to use secure storage?

**Comment by marcosrdz** [comment]
Does the device have more than 100mb of free storage left?

**Comment by ProfLogic** [comment]
All the permissions in the settings menu (only camera and memory are visible) are activated. However, using ADB the result is the following:

`adb shell dumpsys package io.bluewallet.bluewallet | grep granted`

>       android.permission.CHANGE_NETWORK_STATE: granted=true
>       android.permission.FOREGROUND_SERVICE: granted=true
>       android.permission.INTERNET: granted=true
>       android.permission.ACCESS_NETWORK_STATE: granted=true
>       android.permission.USE_FINGERPRINT: granted=true
>       android.permission.VIBRATE: granted=true
>       android.permission.USE_BIOMETRIC: granted=true
>       android.permission.WAKE_LOCK: granted=true
>         android.permission.READ_EXTERNAL_STORAGE: granted=true
>         android.permission.CAMERA: granted=true
>         android.permission.WRITE_EXTERNAL_STORAGE: granted=true

**Comment by ProfLogic** [comment]
The memory settings indicate more than 100 MB of free memory. However, using ADB the result is the following:

`$ adb shell df -H`

> 
> Filesystem            Size  Used Avail Use% Mounted on
> ︙
> /dev/block/mmcblk… 2.1G  1.0G  1.1G  48% /system
> /dev/block/mmcblk…  12G  1.4G   11G  12% /data
> /dev/block/mmcblk… 310M  5.1M  304M   2% /cache
> ︙

**Comment by ProfLogic** [comment]
This bug report is still valid for the current/new version 6.2.3.

**Comment by ProfLogic** [comment]
In addition, after importing a Lightning wallet, the info message “Your wallet has been successfully imported.” is displayed even though the wallet disappears when the App is restarted.

**Comment by marcosrdz** [comment]
Would you be able to record a video of the whole process? This can help us greatly narrow the root cause

**Comment by ProfLogic** [comment]
Attached you find a screen recording which contains:
* starting the app
* showing the splash screen
* displaying the error message
* displaying an empty wallet list
* creating a new Lightning wallet
* displaying the (same) error message (again)
* (not) showing the private/recovery key as QR code (screen recording disabled by the app itself)
* opening the new Lightning wallet
* displaying the initial amount (0 sats)
* closing/terminating the app
* (re)opening the app
* showing the splash screen (again)
* displaying the error message (again)
* displaying an again(!) empty wallet list

https://user-images.githubusercontent.com/89382577/130511269-74ca8007-e02e-456b-9725-a40d82e97322.mp4

**Comment by marcosrdz** [comment]
Ok thank you for the video. This helps us a lot to figure out whats the root cause. I'll update this issue once I have more to share

**Comment by marcosrdz** [comment]
Do you have any security enabled on the device? Such as biometrics or passcode?

**Comment by Overtorment** [comment]
i suppose creating regular onchain wallets throws the same error..?

**Comment by marcosrdz** [comment]
> i suppose creating regular onchain wallets throws the same error..?

Yeah it will cause this is a keystore error.

**Comment by ProfLogic** [comment]
> i suppose creating regular onchain wallets throws the same error..?

That is correct. Also, the error is displayed even before creating a wallet of any type directly after the start of the App.

**Comment by ProfLogic** [comment]
> Do you have any security enabled on the device? Such as biometrics or passcode? Because this error seems to be connected to those features.

Yes, the test was done with a device secured by a four digit PIN code. However, after removing all the security the error and behavior remains the same.

**Comment by marcosrdz** [comment]
I suspect there's underlying differences in the LineageOS Keystore code that makes the standard Keystore code fail. Perhaps a "solution" would be to use Realm in scenarios when this occurs @Overtorment wyd

**Comment by ProfLogic** [comment]
This bug does not occur on three other distinct devices running LineageOS, BlueWallet works well there. Therefore, in general, it is not a problem of the LineageOS keystore implementation.

> I suspect there's underlying differences in the LineageOS Keystore code that makes the standard Keystore code fail. Perhaps a "solution" would be to use Realm in scenarios when this occurs @Overtorment wyd

**Comment by marcosrdz** [comment]
Ok. so this is definitely a good hint as to where the issue lies. what differences in version and settings are there between the devices that succeed and the one that fails?

**Comment by BenGWeeks** [comment]
NB I have also seen this in:

Phone model and OS version: iPhone 12 Pro Max, iOS v15.3.1
BlueWallet app version: v6.2.17

Unfortunately I don't recall exactly what I did before this (although seemed to work OK from memory).

The screenshot I have of the error is on the Wallets home screen with:

```
save to disk exception:
{"message":"error saving key"}
```

Suggest error to user is reformatted as well to:

`Save to disk exception (error saving key)`
